### PR TITLE
Asynchronous configuration / config merging

### LINKED ISSUE #14084: Make it easier to merge configuration into config at runtime
## Context

Related #13952 

We're experimenting with different ways to satisfy security requirements for one of our clients, where they're not allowed to use environment variables to store secrets. We tried a bunch of different approaches; one of them is to load secrets before reading configs and merging them into configuration. This proved surprisingly difficult.  We tried to modify `main()` in `packages/backend/src/index.ts` to read secrets before calling `loadBackendConfig,` but we couldn't figure out the right way to load those values into the config object.

## Feature Suggestion

Provide an API to merge or derive a config object by merging a POJO into a config.

## Possible Implementation

Something like this would be great,

```ts
const static = await loadBackendConfig({
  argv: process.argv,
  logger,
});
  
const config = static.merge({
  backend: {
     database: {
        password: await readFromSecretStore('db_password')
     }
  }
})
```

Is this possible?

### PR DESCRIPTION
**Pull request by RobotSail**
## Hey, I just made a Pull Request!

Closes #14084 

This PR provides the ability to specify a config to `loadBackstageConfig` as an option. This way, values that cannot be stored in a configuration file can be specified and fetched asynchronously during runtime, and merged into the global config before it is consumed by other plugins.

Before this PR, the only way that the config could be provided was through files on the system drive or a remote repository:
```ts
// definition
async function loadBackendConfig(options: {
  remote?: LoadConfigOptionsRemote;
  argv: string[];
}): Promise<{ config: Config }>

// all configs are loaded in here
const { config } = await loadBackendConfig({ argv, remote });
```

With this PR, config objects can now also be provided as an option to `loadBackendConfig`:
```ts
// definition
async function loadBackendConfig(options: {
  remote?: LoadConfigOptionsRemote;
  argv: string[];
  config: JsonObject;
}): Promise<{ config: Config }>

// all configs are loaded in here
const { config } = await loadBackendConfig({ 
  argv,
  remote,
  config: {
    backend: {
      database: {
        connectionURL: await getDatabaseURL()
      }
    }
  }
});
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by RobotSail** [comment]
@taras I wasn't able to extend the ObservableCofigProxy since `setConfig` is a private method, so I simply added `mergeConfig` as a method to the existing class. Let me know what you think of this PR and whether or not this is the approach we want to go for.

File: packages/types/src/json.ts [hunk 1]
```diff
@@ -41,3 +41,62 @@ export interface JsonArray extends Array<JsonValue> {}
...
+ * Attempts to merge two JsonObjects together. In the case of collisions, this function
+ * prefers values from b, unless the value is an object, in which case it recursively
+ * merges the values.
+ *
+ * @param a - The base object
+ * @param b - The object to merge into a
+ * @returns The merged object
+ * @public
+ */
+export const mergeJson = (a: JsonObject, b: JsonObject): JsonObject => {
```
Comment (Rugvip) [inline packages/types/src/json.ts:55]: No need for this actually, the `ConfigReader` has a perfectly good merging implementation already 😁

File: packages/backend-app-api/src/config/config.ts [hunk 2]
```diff
@@ -109,12 +111,12 @@ export async function loadBackendConfig(options: {
       }),
     },
   });
-
   console.info(
     `Loaded config from ${appConfigs.map(c => c.context).join(', ')}`,
   );
 
   config.setConfig(ConfigReader.fromConfigs(appConfigs));
+  config.mergeConfig(options.config ?? {});
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/config.ts:119]: Rather than operating on the completed config reader, I think it's better if we have this happen at the `AppConfig` level instead on the line above. The `AppConfig` represents raw config data along with some context around how it was loaded.

File: packages/backend-app-api/src/config/config.ts [hunk 3]
```diff
@@ -70,6 +71,7 @@ export async function createConfigSecretEnumerator(options: {
 export async function loadBackendConfig(options: {
   remote?: LoadConfigOptionsRemote;
   argv: string[];
+  config?: JsonObject;
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/config.ts:74]: Thinking it's slightly more flexible if we make this `additionalConfig?: AppConfig` instead. It's slightly lower level, but this is a pretty niche customzation point anyway

File: packages/backend-app-api/src/config/ObservableConfigProxy.ts [hunk 4]
```diff
@@ -61,6 +62,12 @@ export class ObservableConfigProxy implements ConfigService {
       },
     };
   }
+  mergeConfig(configData: JsonObject) {
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/ObservableConfigProxy.ts:65]: Let's avoid this addition and merge config data at the `AppConfig` level instead.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍

File: packages/config-loader/src/loader.ts [hunk 5]
```diff
@@ -176,7 +179,7 @@ export async function loadConfig(
     return { fileConfigs, loadedPaths };
   };
 
-  const loadRemoteConfigFiles = async () => {
+  const loadRemoteConfigFiles = async (): Promise<AppConfig[]> => {
```
Comment (Rugvip) [inline packages/config-loader/src/loader.ts:182]: Fine with skipping these changes to `config-loader`? Otherwise we'll need to split this out into a separate changeset as it's just an internal type tweak.

File: packages/backend-app-api/src/config/config.ts [hunk 6]
```diff
@@ -109,11 +111,15 @@ export async function loadBackendConfig(options: {
...
     },
   });
-
   console.info(
     `Loaded config from ${appConfigs.map(c => c.context).join(', ')}`,
   );
 
+  // add the additional config if provided
+  if (options.additionalConfig) {
+    appConfigs.push(options.additionalConfig);
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/config.ts:120]: This needs to be merged into the `newConfigs` in the `onChange` callback above too.

File: packages/backend-app-api/src/config/config.ts [hunk 7]
```diff
@@ -70,6 +71,7 @@ export async function createConfigSecretEnumerator(options: {
 export async function loadBackendConfig(options: {
   remote?: LoadConfigOptionsRemote;
   argv: string[];
+  additionalConfig?: AppConfig;
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/config.ts:74]: I'm feeling we might as well make this `AppConfig[]`, so that you can provide context for config loaded from multiple different external sources.

File: packages/backend-app-api/src/config/ObservableConfigProxy.ts [hunk 8]
```diff
@@ -16,7 +16,7 @@
 
 import { ConfigService } from '@backstage/backend-plugin-api';
 import { ConfigReader } from '@backstage/config';
-import { JsonValue } from '@backstage/types';
+import type { JsonValue } from '@backstage/types';
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/ObservableConfigProxy.ts:19]: No need for this, let's skip it

File: .changeset/spotty-cougars-wink.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/backend-app-api': minor
+'@backstage/config-loader': minor
+---
+
+Introduces the ability to merge configs and JsonObjects
```
Comment (Rugvip) [inline .changeset/spotty-cougars-wink.md:6]: Can drop this one now

File: packages/backend-common/src/config.ts [hunk 10]
```diff
@@ -19,7 +19,7 @@ import {
   loadBackendConfig as newLoadBackendConfig,
 } from '@backstage/backend-app-api';
 import { LoggerService } from '@backstage/backend-plugin-api';
-import { Config } from '@backstage/config';
+import type { AppConfig, Config } from '@backstage/config';
```
Comment (Rugvip) [inline packages/backend-common/src/config.ts:22]: No need for the `type` modifier here

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍 

Couple small tweaks

File: packages/backend-app-api/src/config/config.ts (hunk 6 above)
Comment (RobotSail) [reply packages/backend-app-api/src/config/config.ts:120]: Good catch

File: packages/backend-app-api/src/config/config.ts (hunk 7 above)
Comment (RobotSail) [reply packages/backend-app-api/src/config/config.ts:74]: Good idea.

**Comment by RobotSail** [comment]
@Rugvip Thanks to your help, we managed to cut down this PR size from ~200 LOC added to just 19.

**Comment by taras** [comment]
The result is very nice and simple. Do we need to add this the to documentation in the microsite?

**Comment by RobotSail** [comment]
@Rugvip Did you need any other changes from this PR?

File: .changeset/curly-steaks-mate.md [hunk 11]
```diff
@@ -0,0 +1,7 @@
+---
+'@backstage/backend-app-api': minor
+'@backstage/backend-common': minor
+'@backstage/config-loader': minor
+---
+
+Allow an additionalConfig to be provided to loadBackendConfig that fetches config values during runtime.
```
Comment (Rugvip) [inline .changeset/curly-steaks-mate.md:7]:
```suggestion
---
'@backstage/backend-app-api': minor
'@backstage/backend-common': minor
---

Allow additional configuration data to be provided to `loadBackendConfig` through the new `additionalConfigs` option.
```

`config-loader` is no longer involved, and figure we could message this a bit neater

File: packages/backend-app-api/src/config/config.ts [hunk 12]
```diff
@@ -109,11 +111,13 @@ export async function loadBackendConfig(options: {
       }),
     },
   });
-
   console.info(
     `Loaded config from ${appConfigs.map(c => c.context).join(', ')}`,
   );
 
+  if (options.additionalConfigs) {
+    appConfigs.push(...options.additionalConfigs);
```
Comment (Rugvip) [inline packages/backend-app-api/src/config/config.ts:119]: Question on the usage here, do you think it's best to have this config override any existing `app-config` files, or the other way around? Right now it's overriding config files.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Changeset nit + a bit of discussion

File: packages/backend-app-api/src/config/config.ts (hunk 12 above)
Comment (taras) [reply packages/backend-app-api/src/config/config.ts:119]: I guess what's in `app-config.yaml` should override it to allow developers to hard code secrets while doing local development. What do you think?

Comment (taras) [reply packages/backend-app-api/src/config/config.ts:119]: On second thought, the secret loading code would still run and without proper config, it will fail. I'm not really sure.

Comment (RobotSail) [reply packages/backend-app-api/src/config/config.ts:119]: It should only override configs if the keys collide on non-object types, otherwise to override recursively. 

The way that I had set it up previously with `mergeJson` was that for a given pair of configs such as:
`config A`
```json
{
  "name": "Alex",
  "database": {
    "provider": "mysql",
  },
  "list": [1, 2, 3]
}
```

`config B`
```json
{
  "database": {
    "url": "mysql://user:name@localhost:3206/backstage"
  },
  "list": ["a", "b", "c"]
}
```

Then the result of merged config would look like:
```ts
const mergedConfig = mergeJson(configA, configB);
// mergedConfig:
{
  "name": "Alex",
  "database": {
    "provider": "mysql",
    "url": "mysql://user:name@localhost:3206/backstage"
  },
  "list": ["a", "b", "c"]
}
```

Where the configs further on the right take precedence over those on the left. 

Is this how the `ConfigReader.fromConfigs` function currently works?

Comment (taras) [reply packages/backend-app-api/src/config/config.ts:119]: If I understood @Rugvip correctly, I think the question is "what would the user expect to happen, and how should it work?" The original purpose of this work is to provide an escape hatch for security-conscious adopters who are not allowed to use environment variables to store secrets. Instead of exposing secrets via the environment variable to each Node module in the process, this mechanism will allow us to inject those values into the config without using environment variables. 

But what happens to the configuration in app-config.yaml? Which configuration should win?

Comment (RobotSail) [reply packages/backend-app-api/src/config/config.ts:119]: @Rugvip I guess with my previous comment, all I'm trying to say is that it should only override when there is a non-object collision. Otherwise it should attempt to merge objects recursively. I don't see any reason why the user should have to provide a copy of their config on file just to avoid hard-coding credentials. 

This is outside the scope of the current PR, but maybe we can introduce some form of config schema validation that users can adhere to. This way, if the end result of a config is 
```ts
type Config = {
  database: {
    provider: "mysql" | "sqlite";
    url: string;
  };
}
```

It could error out if it didn't get any of the required values at the end of merging.

Comment (RobotSail) [reply packages/backend-app-api/src/config/config.ts]: @taras The config supplied within code should be expected to be smaller but doesn't have to be. The user should expect the `additionalConfigs` provided to take priority over the configs read from disk or network.

File: .changeset/curly-steaks-mate.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/backend-app-api': minor
+'@backstage/backend-common': minor
```
Comment (Rugvip) [inline .changeset/curly-steaks-mate.md:3]:
```suggestion
'@backstage/backend-app-api': patch
'@backstage/backend-common': patch
```

**Review by Rugvip — APPROVED** [review]
As discussed, let's :shipit: and defer the ordering discussion to future evolution of the config loading 👍 Thank you! 🎉
